### PR TITLE
kubeadm: persist dns resourceRequirements on upgrade

### LINKED ISSUE #3033: kubeadm: keep dns resourceRequirements on upgrade
### What happened?

After an upgrade with kubeadm the resourceRequirements will be reset.
Currently the memory limit is hardcoded to 170M in the template.

In our big prod cluster (1000 nodes,5k services, 20k pods) coreDNS needs more than 170M. 
So we increased the mem limit in the deployment after an outage (all coredns pods got OOM killed)
After kubeadm upgrade the deployment gets updated to 170M 
-> Due to that all coredns pods rotate(during startup the probes are sucessful) and will get oomkilled after loading/caching all dns entries

### What did you expect to happen?

keep resourceRequests after kubeadm upgrade

### How can we reproduce it (as minimally and precisely as possible)?

1. change deployment (increase resourceRequests)
2. run kubeadm upgrade

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by maxl99**
#### What type of PR is this?

/kind bug
/kind feature
#### What this PR does / why we need it:
kubeadm - keep dns resourceRequirements on upgrade

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubeadm/issues/3033

#### Does this PR introduce a user-facing change?

```release-note
keep dns resourceRequirements after kubeadm upgrade
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
i will comment on the issue.

File: cmd/kubeadm/app/phases/addons/dns/dns.go [hunk 1]
```diff
@@ -67,49 +68,71 @@ func DeployedDNSAddon(client clientset.Interface) (string, error) {
...
 }
 
 // EnsureDNSAddon creates the CoreDNS addon
 func EnsureDNSAddon(cfg *kubeadmapi.ClusterConfiguration, client clientset.Interface, out io.Writer, printManifest bool) error {
 	var replicas *int32
+	var resourceRequirements *v1.ResourceRequirements
 	var err error
 	if !printManifest {
-		replicas, err = deployedDNSReplicas(client, coreDNSReplicas)
+		replicas, resourceRequirements, err = deployedDNSDeployment(client, coreDNSReplicas)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns.go:110]: can we just return a Deployment object and take what we need from it?

Comment (maxl99) [reply cmd/kubeadm/app/phases/addons/dns/dns.go:110]: should be possible :) Will update the PR tomorrow

**Comment by neolit123** [comment]
we went for the patches approach, see
https://github.com/kubernetes/kubernetes/pull/124820
/close

**Pull request closed without merging**